File: android/app/src/main/java/com/sugarscratchie/smoke/ui/GameHud.kt
package com.sugarscratchie.smoke.ui

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.viewinterop.AndroidView
import com.airbnb.lottie.LottieComposition
import com.airbnb.lottie.compose.LottieAnimation
import com.airbnb.lottie.compose.LottieConstants

/** Same six targets as the web match bar (first catalog symbols). */
internal val GAME_SYMBOLS = listOf("Heart", "Lock", "Gem", "Star", "Diamond", "Magnet")

@Composable
fun GameTopBar(
    foundMask: Int,
    diamonds: Int,
    coins: Int,
    note: String?,
    symbolCompositions: List<LottieComposition> = emptyList(),
    showWallet: Boolean = true,
    scratchable: Boolean = false,
    onBarCleared: () -> Unit = {},
    modifier: Modifier = Modifier,
) {
    Column(
        modifier = modifier.fillMaxWidth(),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        if (showWallet) {
            Row(
                horizontalArrangement = Arrangement.spacedBy(10.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                WalletChip(mark = "◆", value = diamonds, tint = Color(0xFF7EC8FF))
                WalletChip(mark = "●", value = coins, tint = Color(0xFFFFD56A))
            }
        }
        Box {
            Row(
                modifier =
                    Modifier
                        .clip(RoundedCornerShape(999.dp))
                        .background(Color(0x9E141210))
                        .border(1.dp, Color(0x47FFFFFF), RoundedCornerShape(999.dp))
                        .padding(horizontal = 12.dp, vertical = 8.dp),
                horizontalArrangement = Arrangement.spacedBy(8.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                GAME_SYMBOLS.forEachIndexed { index, name ->
                    SymbolSlot(
                        name = name,
                        found = ((foundMask shr index) and 1) == 1,
                        composition = symbolCompositions.getOrNull(index),
                    )
                }
            }
            if (scratchable) {
                AndroidView(
                    factory = { context ->
                        BarFoilView(context).apply { this.onCleared = onBarCleared }
                    },
                    update = { view -> view.onCleared = onBarCleared },
                    modifier = Modifier.matchParentSize(),
                )
            }
        }
        if (!note.isNullOrBlank()) {
            Text(note, color = Color(0xFFFFE7A3), fontSize = 12.sp)
        }
    }
}

@Composable
private fun WalletChip(mark: String, value: Int, tint: Color) {
    Row(
        modifier =
            Modifier
                .clip(RoundedCornerShape(999.dp))
                .background(Color(0xB3141210))
                .border(1.dp, Color(0x33FFFFFF), RoundedCornerShape(999.dp))
                .padding(horizontal = 10.dp, vertical = 4.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(6.dp),
    ) {
        Text(mark, color = tint, fontSize = 12.sp, fontWeight = FontWeight.Bold)
        Text(value.toString(), color = Color.White, fontSize = 13.sp, fontWeight = FontWeight.SemiBold)
    }
}

@Composable
private fun SymbolSlot(name: String, found: Boolean, composition: LottieComposition?) {
    Box(
        modifier =
            Modifier
                .size(44.dp)
                .clip(CircleShape)
                .background(if (found) Color(0x33FFD278) else Color(0xFA080706))
                .border(
                    1.dp,
                    if (found) Color(0xBFFFDC8C) else Color(0x33FFFFFF),
                    CircleShape,
                ),
        contentAlignment = Alignment.Center,
    ) {
        if (composition != null) {
            LottieAnimation(
                composition = composition,
                iterations = LottieConstants.IterateForever,
                modifier = Modifier.size(32.dp).alpha(if (found) 1f else 0.45f),
            )
        } else {
            val tint = if (found) Color(0xFFFFE7A3) else Color(0xFF6A6560)
            SymbolMark(name = name, color = tint, modifier = Modifier.size(26.dp))
        }
    }
}

@Composable
fun SymbolMark(name: String, color: Color, modifier: Modifier = Modifier) {
    Canvas(modifier) {
        val w = size.width
        val h = size.height
        val stroke = Stroke(width = w * 0.09f)
        when (name) {
            "Heart" -> {
                val path =
                    Path().apply {
                        moveTo(w * 0.5f, h * 0.88f)
                        cubicTo(w * -0.05f, h * 0.45f, w * 0.18f, h * 0.05f, w * 0.5f, h * 0.32f)
                        cubicTo(w * 0.82f, h * 0.05f, w * 1.05f, h * 0.45f, w * 0.5f, h * 0.88f)
                        close()
                    }
                drawPath(path, color)
            }
            "Lock" -> {
                drawRoundRect(
                    color,
                    topLeft = Offset(w * 0.18f, h * 0.42f),
                    size = Size(w * 0.64f, h * 0.48f),
                    cornerRadius = androidx.compose.ui.geometry.CornerRadius(w * 0.08f),
                )
                drawArc(
                    color,
                    startAngle = 200f,
                    sweepAngle = 140f,
                    useCenter = false,
                    topLeft = Offset(w * 0.28f, h * 0.08f),
                    size = Size(w * 0.44f, h * 0.48f),
                    style = stroke,
                )
            }
            "Gem" -> {
                val path =
                    Path().apply {
                        moveTo(w * 0.5f, h * 0.08f)
                        lineTo(w * 0.88f, h * 0.38f)
                        lineTo(w * 0.5f, h * 0.92f)
                        lineTo(w * 0.12f, h * 0.38f)
                        close()
                    }
                drawPath(path, color)
            }
            "Star" -> {
                val path = starPath(w, h)
                drawPath(path, color)
            }
            "Diamond" -> {
                val path =
                    Path().apply {
                        moveTo(w * 0.5f, h * 0.06f)
                        lineTo(w * 0.92f, h * 0.5f)
                        lineTo(w * 0.5f, h * 0.94f)
                        lineTo(w * 0.08f, h * 0.5f)
                        close()
                    }
                drawPath(path, color, style = stroke)
            }
            else -> {
                drawCircle(color, radius = w * 0.18f, center = Offset(w * 0.28f, h * 0.42f))
                drawCircle(color, radius = w * 0.18f, center = Offset(w * 0.72f, h * 0.42f))
                drawLine(color, Offset(w * 0.46f, h * 0.42f), Offset(w * 0.54f, h * 0.42f), strokeWidth = w * 0.08f)
                drawLine(color, Offset(w * 0.5f, h * 0.42f), Offset(w * 0.5f, h * 0.86f), strokeWidth = w * 0.08f)
            }
        }
    }
}

private fun starPath(w: Float, h: Float): Path {
    val cx = w * 0.5f
    val cy = h * 0.52f
    val outer = w * 0.46f
    val inner = w * 0.2f
    return Path().apply {
        for (i in 0 until 10) {
            val radius = if (i % 2 == 0) outer else inner
            val angle = Math.toRadians((-90.0 + i * 36.0))
            val x = cx + (radius * kotlin.math.cos(angle)).toFloat()
            val y = cy + (radius * kotlin.math.sin(angle)).toFloat()
            if (i == 0) moveTo(x, y) else lineTo(x, y)
        }
        close()
    }
}

@Composable
fun BodySymbol(name: String, modifier: Modifier = Modifier) {
    Box(
        modifier =
            modifier
                .size(56.dp)
                .offset(x = (-28).dp, y = (-28).dp)
                .clip(CircleShape)
                .background(Color(0xE6141210))
                .border(1.dp, Color(0xBFFFDC8C), CircleShape),
        contentAlignment = Alignment.Center,
    ) {
        SymbolMark(name = name, color = Color(0xFFFFE7A3), modifier = Modifier.size(32.dp))
    }
}

File: ios/SugarScratchie/UI/GameHud.swift
import Lottie
import SwiftUI

let gameSymbols = ["Heart", "Lock", "Gem", "Star", "Diamond", "Magnet"]

struct GameTopBar: View {
    var foundMask: Int
    var diamonds: Int
    var coins: Int
    var note: String?
    var symbolAnimations: [LottieAnimation] = []
    var showWallet: Bool = true
    var scratchable: Bool = false
    var onBarCleared: () -> Void = {}

    var body: some View {
        VStack(spacing: 8) {
            if showWallet {
                HStack(spacing: 10) {
                    WalletChip(mark: "◆", value: diamonds, tint: Color(red: 0.494, green: 0.784, blue: 1))
                    WalletChip(mark: "●", value: coins, tint: Color(red: 1, green: 0.835, blue: 0.416))
                }
            }
            ZStack {
                HStack(spacing: 8) {
                    ForEach(Array(gameSymbols.enumerated()), id: \.offset) { index, name in
                        SymbolSlot(
                            name: name,
                            found: ((foundMask >> index) & 1) == 1,
                            animation: symbolAnimations.indices.contains(index) ? symbolAnimations[index] : nil
                        )
                    }
                }
                .padding(.horizontal, 12)
                .padding(.vertical, 8)
                .background(Color(red: 0.078, green: 0.071, blue: 0.063).opacity(0.62))
                .overlay(
                    RoundedRectangle(cornerRadius: 999)
                        .stroke(Color.white.opacity(0.28), lineWidth: 1)
                )
                .clipShape(RoundedRectangle(cornerRadius: 999))

                if scratchable {
                    BarFoilView(onCleared: onBarCleared)
                        .frame(maxWidth: .infinity, maxHeight: .infinity)
                }
            }
            if let note, !note.isEmpty {
                Text(note)
                    .font(.system(size: 12))
                    .foregroundStyle(Color(red: 1, green: 0.906, blue: 0.639))
            }
        }
        .frame(maxWidth: .infinity)
    }
}

private struct WalletChip: View {
    var mark: String
    var value: Int
    var tint: Color

    var body: some View {
        HStack(spacing: 6) {
            Text(mark)
                .font(.system(size: 12, weight: .bold))
                .foregroundStyle(tint)
            Text("\(value)")
                .font(.system(size: 13, weight: .semibold))
                .foregroundStyle(.white)
        }
        .padding(.horizontal, 10)
        .padding(.vertical, 4)
        .background(Color(red: 0.075, green: 0.071, blue: 0.063).opacity(0.7))
        .overlay(
            RoundedRectangle(cornerRadius: 999)
                .stroke(Color.white.opacity(0.2), lineWidth: 1)
        )
        .clipShape(RoundedRectangle(cornerRadius: 999))
    }
}

private struct SymbolSlot: View {
    var name: String
    var found: Bool
    var animation: LottieAnimation?

    var body: some View {
        ZStack {
            Circle()
                .fill(found ? Color(red: 1, green: 0.824, blue: 0.471).opacity(0.2) : Color(red: 0.031, green: 0.027, blue: 0.024).opacity(0.98))
            Circle()
                .stroke(found ? Color(red: 1, green: 0.863, blue: 0.549).opacity(0.75) : Color.white.opacity(0.2), lineWidth: 1)
            if let animation {
                LottieView(animation: animation)
                    .playing(loopMode: .loop)
                    .frame(width: 32, height: 32)
                    .opacity(found ? 1 : 0.45)
            } else {
                SymbolMark(name: name, color: found ? Color(red: 1, green: 0.906, blue: 0.639) : Color(red: 0.416, green: 0.396, blue: 0.376))
                    .frame(width: 26, height: 26)
            }
        }
        .frame(width: 44, height: 44)
    }
}

struct SymbolMark: View {
    var name: String
    var color: Color

    var body: some View {
        Canvas { context, size in
            let w = size.width
            let h = size.height
            switch name {
            case "Heart":
                var path = Path()
                path.move(to: CGPoint(x: w * 0.5, y: h * 0.88))
                path.addCurve(
                    to: CGPoint(x: w * 0.5, y: h * 0.32),
                    control1: CGPoint(x: w * -0.05, y: h * 0.45),
                    control2: CGPoint(x: w * 0.18, y: h * 0.05)
                )
                path.addCurve(
                    to: CGPoint(x: w * 0.5, y: h * 0.88),
                    control1: CGPoint(x: w * 0.82, y: h * 0.05),
                    control2: CGPoint(x: w * 1.05, y: h * 0.45)
                )
                path.closeSubpath()
                context.fill(path, with: .color(color))
            case "Lock":
                context.fill(
                    Path(roundedRect: CGRect(x: w * 0.18, y: h * 0.42, width: w * 0.64, height: h * 0.48), cornerRadius: w * 0.08),
                    with: .color(color)
                )
                var arc = Path()
                arc.addArc(
                    center: CGPoint(x: w * 0.5, y: h * 0.32),
                    radius: w * 0.22,
                    startAngle: .degrees(200),
                    endAngle: .degrees(340),
                    clockwise: false
                )
                context.stroke(arc, with: .color(color), lineWidth: w * 0.09)
            case "Gem":
                var path = Path()
                path.move(to: CGPoint(x: w * 0.5, y: h * 0.08))
                path.addLine(to: CGPoint(x: w * 0.88, y: h * 0.38))
                path.addLine(to: CGPoint(x: w * 0.5, y: h * 0.92))
                path.addLine(to: CGPoint(x: w * 0.12, y: h * 0.38))
                path.closeSubpath()
                context.fill(path, with: .color(color))
            case "Star":
                context.fill(starPath(w: w, h: h), with: .color(color))
            case "Diamond":
                var path = Path()
                path.move(to: CGPoint(x: w * 0.5, y: h * 0.06))
                path.addLine(to: CGPoint(x: w * 0.92, y: h * 0.5))
                path.addLine(to: CGPoint(x: w * 0.5, y: h * 0.94))
                path.addLine(to: CGPoint(x: w * 0.08, y: h * 0.5))
                path.closeSubpath()
                context.stroke(path, with: .color(color), lineWidth: w * 0.09)
            default:
                context.fill(Path(ellipseIn: CGRect(x: w * 0.1, y: h * 0.24, width: w * 0.36, height: h * 0.36)), with: .color(color))
                context.fill(Path(ellipseIn: CGRect(x: w * 0.54, y: h * 0.24, width: w * 0.36, height: h * 0.36)), with: .color(color))
                var stem = Path()
                stem.move(to: CGPoint(x: w * 0.46, y: h * 0.42))
                stem.addLine(to: CGPoint(x: w * 0.54, y: h * 0.42))
                stem.move(to: CGPoint(x: w * 0.5, y: h * 0.42))
                stem.addLine(to: CGPoint(x: w * 0.5, y: h * 0.86))
                context.stroke(stem, with: .color(color), lineWidth: w * 0.08)
            }
        }
    }

    private func starPath(w: CGFloat, h: CGFloat) -> Path {
        let cx = w * 0.5
        let cy = h * 0.52
        let outer = w * 0.46
        let inner = w * 0.2
        var path = Path()
        for i in 0..<10 {
            let radius = i % 2 == 0 ? outer : inner
            let angle = (-90.0 + Double(i) * 36.0) * Double.pi / 180
            let x = cx + radius * CGFloat(cos(angle))
            let y = cy + radius * CGFloat(sin(angle))
            if i == 0 { path.move(to: CGPoint(x: x, y: y)) }
            else { path.addLine(to: CGPoint(x: x, y: y)) }
        }
        path.closeSubpath()
        return path
    }
}
